src/js/mixins/stateful.js

```diff
@@ -0,0 +1,92 @@
+/**
+ * @file mixins/stateful.js
+ */
+import * as Fn from '../utils/fn';
+import log from '../utils/log';
+import * as Obj from '../utils/obj';
+
+/**
+ * Set the state of an object by mutating its `state` object in place.
+ *
+ * @param  {Object|Function} stateUpdates
+ *         A new set of properties to shallow-merge into the plugin state. Can
+ *         be a plain object or a function returning a plain object.
+ *
+ * @return {Object}
+ *         An object containing changes that occurred. If no changes occurred,
+ *         returns `undefined`.
+ */
+const setState = function(stateUpdates) {
+
+  // Support providing the `stateUpdates` state as a function.
+  if (typeof stateUpdates === 'function') {
+    stateUpdates = stateUpdates();
+  }
+
+  if (!Obj.isPlain(stateUpdates)) {
+    log.warn('non-plain object passed to `setState`', stateUpdates);
```

Comment:
Also, what about just going with it regardless of what kind of object was passed in, assuming it isn't `null`. We could copy the properties from whatever was passed in.